packages/devtools-map-bindings/src/remapScopes.js

```diff
@@ -0,0 +1,96 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// @flow
+
+import type {
+  Scope,
+  SourceScope,
+  MappedScopeBindings,
+  SyntheticScope
+} from "debugger-html";
+
+export type RemappedScope = {
+  scopes: SyntheticScope[],
+  start: number,
+  end: number
+};
+
+// Performs mapping of the original parsed scopes to the locals mappings
+// based on the generated source parse and source map data.
+function remapScopes(
+  scopes: ?(SourceScope[]),
+  generatedScopes: MappedScopeBindings[]
+): ?(RemappedScope[]) {
+  if (!scopes || scopes.length === 0) {
+    return null;
+  }
+  // Chunk original source scopes on function/closure boundary
+  const { result: scopeChunks } = scopes.reduce(
+    ({ isLast, result }, scope) => {
+      if (isLast) {
+        result.push([]);
+      }
+      result[result.length - 1].push(scope);
+      return {
+        isLast: scope.type === "function",
+        result
+      };
+    },
+    { isLast: true, result: [] }
+  );
+  const { result: assigned } = scopeChunks.reduce(
+    ({ result, searchIn, searchOffset }, scopeChunk) => {
+      if (searchIn.length === 0) {
+        return { result, searchIn, searchOffset };
+      }
+      // Process chunk of original parsed scopes: create used original names
+      // binding summary per scope and entire chunk.
+      const summarizedScopes = scopeChunk.map(({ type, bindings }) => ({
+        type,
+        bindingsNames: Object.keys(bindings)
+      }));
+      const names = summarizedScopes.reduce(
+        (acc, { bindingsNames }) => acc.concat(bindingsNames),
+        []
+      );
```

Comment:
it would be nice to have a small helper function: `getBindingNames` that can take our function scopes and return a list of names...

```js
const names = getBindingNames(scopeChunk)
```